src/test/fuzz/base_encode_decode.cpp

```diff
@@ -6,49 +6,90 @@
 
 #include <base58.h>
 #include <psbt.h>
+#include <test/fuzz/FuzzedDataProvider.h>
 #include <util/strencodings.h>
 #include <util/string.h>
 
 #include <cassert>
-#include <cstdint>
 #include <string>
 #include <vector>
+#include <ranges>
 
-using util::TrimString;
 using util::TrimStringView;
 
-FUZZ_TARGET(base_encode_decode)
+FUZZ_TARGET(base58_encode_decode)
 {
-    const std::string random_encoded_string(buffer.begin(), buffer.end());
+    FuzzedDataProvider provider(buffer.data(), buffer.size());
+    const std::string random_string{provider.ConsumeRemainingBytesAsString()};
```

Comment:
I'm afraid `FuzzedDataProvider::remaining_bytes_` becomes zero after this line in the current version, meaning `provider.ConsumeIntegralInRange<int>(0, decoded.size() - 1)` below will always return the minimum value if I'm reading the `ConsumeIntegralInRange`-implementation correctly.

```suggestion
    const std::string random_string{provider.ConsumeRandomLengthString(101)};
```
---

Could maybe (in another PR) add something like `bool FuzzedDataProvider::consume_remaining_called_{false}` which is set by `ConsumeRemainingBytes*()` and then assert against it in the various consume methods, to catch cases like these.

Reply:
[Done](https://github.com/bitcoin/bitcoin/compare/5bc0a8daab37a826a59ed83eb59aef4081f1fe19..6fd185c035c1cc4dd961cf14a2087e97fb069440), thanks

Reply:
@dergoegge any thoughts on my suggestion of altering `FuzzedDataProvider` to trigger asserts when calls like `ConsumeIntegralInRange<int>(0, decoded.size() - 1)` are made after calls to `ConsumeRemainingBytes*()`? Maybe such an improvement could be upstreamed even (possibly with a flag to turn it off for sloppy tests).

Reply:
It might add extra friction to agree on preprocessor logic to make the code still compile on other compilers. Also, doing it at compile-time means it's an all/nothing thing that can't be turned off for individual sloppy tests. But I agree it would be great to see if upstream would take a compile-time version first.

Reply:
Yeah, hopefully shouldn't be much of an obstacle, maybe I'm too pessimistic about that aspect.

Reply:
Is there anything else for me to do here?

Reply:
Nothing more to do as far as this PR goes. Still interested to hear from folks more focused on fuzzing.

Reply:
Re https://github.com/bitcoin/bitcoin/pull/30746#discussion_r1824164226: I think I agree with marco that a compile time check would be nicer and we should ideally upstream it so that our `FuzzedDataProvider` impl stays in sync.

Reply:
Great that you made time to look into this!

> An alternative could be to forbid the direct use of the depleting member functions (with a linter) and add a new helper which eats a `FuzzedDataProvider&&` and returns the bytes in it. This way, the existing use-after-move checks can be used.

So you mean disallowing direct calls to `ConsumeRemainingBytes*`-family of methods only, and provide a helper that does only that part instead after swallowing up `FuzzedDataProvider&&` so it should no longer be used. Sounds like it's worth a try to me!